File: game.cpp
#include <iostream>
using namespace std;

void configuration(); //The Game's Configuration Options Display at Start
void BoardSize(); //To define the Board-Size
void Difficulty(); //To define the Difficulty of the Game
void Player(); //To Assign the Player(s) Marks

void Game(); //The Main Game-Sequence
void turn(); //Player's Turn
void Board(); //Print the Whole Board
int over(); //Check if anybody Won and Completed the Konnect-Goal
int draw(); //Check if the Board fills-out
void PlayAgain(); //The PlayAgain Option

char configure = 'N'; //The Default Configuration Options 
int max_cIgnore = 255; //MaxCharacters to be Ignored in the Buffer
char playAgain = 'Y'; //The Default PlayAgain Option

char board[30][30]; //The Game Board
int rows = 7, cols = 7 ; //Total Default Rows and Columns of the Board
int r,c; //For Loop-controls Row & Column

int konnect = 4; //The Default Konnect-Number Goal of the Game
int k; // For Loop-control Konnect

char emp = '-'; //Empty Board's Mark
char player[8 +1] = {emp,'X','O','I','H','E','V','Z','M'}; //The Player's Default Marks
int players = 2 +1; //Total Default Number of Players + 1
int p; //For Loop-control and also Player Number

int a,b; //The Player Inputs, for Specific Row & Column

int main()
{	
	cout << endl
		<< "Welcome to the Konnect-Board-Game. Its a Two-Player Turn-based Game. \n"
		<< "You are Advised to Consult the Readme File, before Playing the Game. \n"
		<< endl;

	configuration();
	p = 1; //Default. First-Turn for Player 1
	
	do {	
		cout << "So, Lets Begin! \n\n";
		cout << "Pick the Specified Place according to the Index of the Row and Column. \n\n";
		
		//Initialize the whole Board
		for (r=0; r<rows; r++) {
			cout << "\t";
			for (c=0; c<cols; c++) {
				board[r][c] = emp;
				cout << board[r][c] << " ";
			}
			cout << endl;
		}
		cout << endl;
	
		Game();
		PlayAgain();
	}
	while(playAgain == 'Y');
	
	cout << endl << endl;
	return 0;
}

void configuration()
{
	do {	
		if(configure == 'Y') {
			BoardSize();
			Difficulty();
			Player();
		}
		
		cout << "The Board-Size is (" << rows << "x" << cols << "). \n"
			<< "The Konnect-Goal is to Occupy " << konnect << " Consecutive Places. \n";
		
		for(p=1; p<players; p++)
			cout << " The Player " << p << "'s Mark is '" << player[p] << "'. \n";
		
		cout << "   Do you want to Configure Anything (Y/N): ";
		
		cin >> configure;
		configure = toupper(configure);
		cin.ignore(max_cIgnore, '\n');
		
		cout << endl << endl;
	}
	while(configure == 'Y');
}

void BoardSize()
{
	char board_size;
	int size_sm = 6, size_md = 8, size_lg = 10; //Predefined Sizes
	
	cout << "Choose the Board-Size (s:small, m:medium, l:large, c:custom): ";
	cin >> board_size;
	board_size = toupper(board_size);
	cin.ignore(max_cIgnore, '\n');

	switch (board_size) 
	{
		case 'S':
			rows = size_sm, cols = size_sm; 
			break;
		case 'M':
			rows = size_md, cols = size_md; 
			break;
		case 'L':
			rows = size_lg, cols = size_lg;
			break;
		
		case 'C':
			cout << "   Enter the Custom Total Rows and Columns of the Board: ";
			cin >> rows >> cols;
			cin.ignore(max_cIgnore,'\n');
			
			if (cin.fail()) {
				cin.clear(); cin.ignore(max_cIgnore,'\n');
				cout << "   Oops! Its not the legal Row-Column Numbers. \n\n";
				cout << "Again, ";
				BoardSize();
			}
			else if (!(rows>=2 && cols>=2)) {
				cout << "   Oops! It's too Small to be a Board. \n\n";
				cout << "Again, ";
				BoardSize();
			}
			else if (!(rows<=30 && cols<=30)) {
				cout << "   Oops! It's too Large to be a Board. \n\n";
				cout << "Again, ";
				BoardSize();
			}
			break;
		
		default:
			cout << "The Board-Size is Set to Default 7x7. \n";
			rows = 7, cols = 7;
	}
	cout << endl; //It would be Buffered each Time the Function Self-Calls
}

void Difficulty()
{
	char difficulty; //Default
	int easy = 3, inter = 5, hard = 6; //Predefined Difficulties
	
	cout << "Choose the Game-Difficulty (e:easy, i:intermediate, h:hard, c:custom): ";
	cin >> difficulty;
	difficulty = toupper(difficulty);
	cin.ignore(max_cIgnore, '\n');

	switch (difficulty) 
	{
		case 'E':
			konnect = easy;
			break;
		case 'I':
			konnect = inter;
			break;
		case 'H':
			konnect = hard;
			break;
		
		case 'C':
			cout << "   Enter the Custom Konnect-Number Goal of the Game: ";
			cin >> konnect;
			cin.ignore(max_cIgnore,'\n');
			
			if (cin.fail()) {
				cin.clear(); cin.ignore(max_cIgnore, '\n');
				cout << "   Oops! Its not the legal Konnect-Goal Number. \n\n";
				cout << "Again, ";
				Difficulty();
			}
			else if (!(konnect>=2)) {
				cout << "   Oops! It's too Easy to be a Game. \n\n";
				cout << "Again, ";
				Difficulty();
			}
			else if (!(konnect<=12)) {
				cout << "   Oops! It's too Difficult to be a Game. \n\n";
				cout << "Again, ";
				Difficulty();
			}
			break;
		
		default:
			cout << "The Konnect-Goal Difficulty is Set to Default 4. \n";
			konnect = 4;
	}

	if(konnect > rows && konnect > cols) {
		cout << "Oops! It isn't Possible to Complete this Goal on that Small Board Size. \n";
		cout << "Choose the Board-Size Again! \n\n";
		configuration();
	}
	
	cout << endl; //It would be Buffered each Time Function Self-Calls
}

void Player()
{
	do {
		cout << "How Many Players are there to Play: ";
		cin >> players;
		cin.ignore(max_cIgnore, '\n');
		players++;

		if (cin.fail()) {
			cin.clear();cin.ignore(max_cIgnore,'\n');
			cout << "It's Set to Default 2-Player. \n";
			players = 2 +1;
		}
		else if (!(players-1>=2)) {
			cout << "   Oops! It's too Less Players to Play a Game. \n\n";
			cout << "Again, ";
			continue;
		}
		else if (!(players-1<=8)) {
			cout << "   Oops! It's too Many Players. \n\n";
			cout << "Again, ";
			continue;
		}
		else 
			break;
	}
	while(true);
	
	cout << endl;

	char customMark; //Default
	cout << "Do you Want to Customize the Player's Marks (Y/N): ";
	cin >> customMark;
	customMark = toupper(customMark);
	cin.ignore(max_cIgnore, '\n');
	
	if(customMark == 'Y') {
		for(p=1; p<players; p++) {
			cout << "   Enter the Custom Identification-Mark for Player " << p << ": ";
			cin >> player[p]; //player[0] is Empty
			cin.ignore(max_cIgnore, '\n');
		}
	}
	else
		cout << "The Player Marks are Set to Default. \n";
	
	cout << endl;
}


void Game()
{
	while (true) {
		
		if(!(p <= players-1)) {
			p = 1;
		}
		
		cout << "Player " << p << "'s Turn: ";
		turn();
		
		if (over()) {
			cout << "Player " << p << " Won! \n\n";
			break;
		}
		if (draw()) {
			cout << "Its a Draw! \n\n";
			break;
		}

		p++;
	}
}

void turn()
{
	cin >> a >> b;
	cin.ignore(max_cIgnore,'\n');
			
	if (cin.fail()) {
		cin.clear(); cin.ignore(max_cIgnore, '\n');
		cout << "Oops! Its not Legal. \n"
			<< "Pick Again, (Rows 1 to " << rows << ") & "
						<< "(Columns 1 to " << cols << "): ";
		turn();
	}
	else if (a>=1 && a<=rows && b>=1 && b<=cols) {
		
		if (board[a-1][b-1] == emp) {
			board[a-1][b-1] = player[p];
			Board();
		}
		else {
			cout << "Its already Occupied, Pick Somewhere else: ";
			turn();
		}
	}
	else {
		cout << "Oops! Its Outside the Board. \n"
			<< "Pick Again, (Rows 1 to " << rows << ") & "
						<< "(Columns 1 to " << cols << "): ";
		turn();
	}
	// There's No Buffered Output if the Function Self-Calls
}

void Board()
{
	cout << endl;
	for (r=0; r<rows; r++) {
		cout << "\t";
		for (c=0; c<cols; c++)
			cout << board[r][c] << " ";
		cout << endl;
	}
	cout << endl << endl;
}

int over()
{
	for (r=0; r<rows; r++) {
		for (c=0; c<cols; c++) {
			
			for(k=0; k<=konnect; k++)
				if (board[r][c+k] != player[p])
					break;
			if(k==konnect)
				return true; //Horizonatlly
			
			for(k=0; k<=konnect; k++)
				if (board[r+k][c] != player[p])
					break;
			if(k==konnect)
				return true; //Vertically
			
			for(k=0; k<=konnect; k++)
				if (board[r+k][c+k] != player[p])
					break;
			if(k==konnect)
				return true; //Diagonally-Right

			for(k=0; k<=konnect; k++)
				if (board[r+k][c-k] != player[p])
					break;
			if(k==konnect)
				return true; //Diagonally-Left
		}
	}
	
	return false;
}

int draw()
{
	for (r=0; r<rows; r++)
		for (c=0; c<cols; c++)
			if (board[r][c] == emp) //false if even one space is empty
				return false;
	return true;
}

void PlayAgain()
{
	cout << "Do You Want to Play Again (Y/N): ";
	cin >> playAgain;
	playAgain = toupper(playAgain);
	cin.ignore(max_cIgnore, '\n');

	if(playAgain == 'Y') {
		cout << "Do You Want to Configure the Options (Y/N): ";	
		cin >> configure;
		configure = toupper(configure);
		cin.ignore(max_cIgnore, '\n');
		
		cout << endl << endl;
		if(configure == 'Y')
			configuration();
		
		p = (rand() % (players-1)) + 1; //Next Time, the Random Player goes for the First Turn
	}
	else if(playAgain == 'N') {
		cout << endl << endl;
		cout << "Thanks for Playing. Hope You Enjoyed the Game! \n";
		cout << "It would be very Kind of you, if you give us an Honest Feedback. \n\n";

		cout << "You can Find me in GitHub https://github.com/Haseebullah9012. \n";
	}
	else {
		cout << "   Oops! Its not a legal Response. \n\n";
		cout << "Again, ";
		PlayAgain();
	}
}

File: game.js

/* In Development */
// Player Symbol Customization Feature
// Details onExit 
// Predefined Game Modes


//The Game Board
var board = []; 
for(r=0; r<12; r++)
    board[r] = []; 

var rows = document.getElementById("Rows").value = 4; //Total Default Rows of the Board
var cols = document.getElementById("Cols").value = 4; //Total Default Columns of the Board
var konnect = document.getElementById("Konnect").value = 3; //The Default Konnect-Number Goal of the Game
var players = document.getElementById("Players").value = 2; //Total Default Number of Players

var r,c; //For Loop-controls Row & Column
var k; // For Loop-control Konnect
var p; //For Loop-control and also Player Number

var emp = ""; //Empty Board's Mark
var player = [emp,'X','O','I','H','E','V','Z','M']; //The Player's Default Marks

var validated; // The FormValidation Condition
var a,b; //The Player Inputs


//The HTMl Elements
var Docs = document.getElementById("docs"); //The Docs Container
var configure = document.getElementById("Configure"); //The Configuration Options Conatiner
var play = document.getElementById("play"); //The Play Button
var newGame = document.getElementById("newGame"); //The New Game Button
var Info = document.getElementById("info"); //The Player Info Text
var GameContainer = document.getElementById("game-container"); //The Board-Game Container

var Speaker = document.getElementById("SpeakerIcon"); //Background Music is Playing
var SpeakerMuted = document.getElementById("SpeakerMuted-Icon"); //Background Music is Muted
var Click = document.getElementById("Square-Click"); // The Button-CLick
var MenuClick = document.getElementById("Menu-Click"); // The Menu Button-CLick
var GameMusic = document.getElementById("Game-Music"); // The Background Game-Music
var ConfigureMusic = document.getElementById("Configure-Music"); // The Background Configuration- Music


//Starting Element States
newGame.style.display = "none";
Info.style.display = "none";
GameContainer.style.display = "none";
Speaker.style.display = "none";
GameMusic.muted = true;
ConfigureMusic.muted = true;


function Play()
{
    MenuClick.pause();
    MenuClick.currentTime = 0; //To Stop if Previous CLick is still Running
    MenuClick.play();

    configuration();
    if(validated == true) {
        Docs.style.display = "none";
        configure.style.display = "none";
        play.style.display = "none";
        
        newGame.style.display = "block";
        Info.style.display = "block";
        printBoard();

        GameMusic.play();
        ConfigureMusic.pause();
        ConfigureMusic.currentTime = 0; //To Stop Other Music
        
        /* Because the Music Don't Load on Startup */
        if(GameMusic.muted == false && ConfigureMusic.muted == false) {
            SpeakerMuted.style.display = "none";
            Speaker.style.display = "block"; 
        }
    }
}

function NewGame()
{
    MenuClick.pause();
    MenuClick.currentTime = 0; //To Stop if Previous CLick is still Running
    MenuClick.play();

    resetBoard();
    
    Docs.style.display = "block";
    configure.style.display = "block";
    play.style.display = "block";
    
    newGame.style.display = "none";
    Info.style.display = "none";
    
    ConfigureMusic.play();
    GameMusic.pause();
    GameMusic.currentTime = 0; //To Stop Other Music
}


function resetBoard() {
    GameContainer.classList.remove("end");
    GameContainer.style.display = "none";
    
    Info.classList.remove("large");
    newGame.classList.remove("big");
    
    while(GameContainer.hasChildNodes())
        for(r=0; r<rows; r++)
            for(c=0; c<cols; c++)
                GameContainer.removeChild(board[r][c]);
}

function configuration() {
    rows = document.getElementById("Rows").value;
    cols = document.getElementById("Cols").value;
    konnect = document.getElementById("Konnect").value;
    players = document.getElementById("Players").value;
    p = 1; //Default First Turn

    let RowsValid = true;
    let ColsValid = true;
    let KonnectValid = true;
    let PlayersValid = true;
        
    if(isNaN(rows) || !(rows>=2 && rows<=12)) {
        RowsValid = false;
        document.getElementById("Rows").classList.add("invalid");
    }
    else {
        RowsValid = true;
        document.getElementById("Rows").classList.remove("invalid");
    }
    
    if(isNaN(cols) || !(cols>=2 && cols<=12)) {
        ColsValid = false;
        document.getElementById("Cols").classList.add("invalid");
    }
    else {
        ColsValid = true;
        document.getElementById("Cols").classList.remove("invalid");
    }
    
    if(isNaN(konnect) || !(konnect>=2 && konnect<=8)) {
        KonnectValid = false;
        document.getElementById("Konnect").classList.add("invalid");
    }
    else if(!(konnect<=rows || konnect<=cols) && !(rows>=8 || cols>=8)) {
        KonnectValid = false;
        document.getElementById("Konnect").classList.add("invalid");
    }
    else {
        KonnectValid = true;
        document.getElementById("Konnect").classList.remove("invalid");
    }

    if(isNaN(players) || !(players>=2 && players<=8)) {
        PlayersValid = false;
        document.getElementById("Players").classList.add("invalid");
    }
    else {
        PlayersValid = true;
        players++; //Because the Index 0 of player Symbols-Array is Empty
        document.getElementById("Players").classList.remove("invalid");
    } 
    
    if(RowsValid==true && ColsValid==true && KonnectValid== true && PlayersValid==true)
        validated = true;
    else
        validated = false;
}

function printBoard() {

    Info.textContent = "Player " + p + "'s Turn";
    GameContainer.style.display = "block";

    for(r=0; r<rows; r++) {
        for(c=0; c<cols; c++) {
            board[r][c] = document.createElement('button');
            board[r][c].textContent = emp;
            board[r][c].dataset.x = r, board[r][c].dataset.y = c;
            board[r][c].classList.add("square");
            board[r][c].addEventListener("click", turn);
            
            GameContainer.appendChild(board[r][c]);
        }
        //GameContainer.append(" ");
    }
    
    setBoardDimensions();
}

function setBoardDimensions() {
    
    var Board_width, Board_height; //For Board Dimensions
    var Square; //For Square Dimensions
    
    if(rows==cols) {
        if(rows==2)
            Square = 160/rows;
        else if(rows==3)
            Square = 250/rows;
        else if(rows==4)
            Square = 320/rows;
        else
            Square = 520/rows;
    }
    else if(rows<10 && cols<10) {
        if(rows>cols) {
            Square = 480/rows;
        }
        else {
            Square = 680/cols;
        }
    }
    else {
        if(rows>cols) {
            if(rows==2)
                Square = 200/rows;
            else if(rows==3)
                Square = 280/rows;
            else if(rows==4)
                Square = 360/rows;
            else
                Square = 440/rows;
        }
        else {
            Square = 560/rows;
        }
    }
    
    Board_width = Square*cols + 5;
    Board_height = Square*rows + 3;

    GameContainer.style.width = Board_width + "px";
    GameContainer.style.height = Board_height + "px";
    
    for(r=0; r<rows; r++) {
        for(c=0; c<cols; c++) {
            board[r][c].style.width = Square + "px";
            board[r][c].style.height = Square + "px";
        }
    }
}


//On User Inpput
function turn() {
    Click.pause();
    Click.currentTime = 0; //To Stop if Previous Click is still Running

    a = parseInt($(this).data('x'));
    b = parseInt($(this).data('y'));
    
    if(board[a][b].textContent == emp) {
        board[a][b].textContent = player[p];
        Click.play();
        
        if(over() == true) {
            Info.textContent = "Player " + p + " Won!";
            Info.classList.add("large");
            GameContainer.classList.add("end");
            newGame.classList.add("big");
            return;
        }
        
        if(draw() == true) {
            Info.textContent = "Its a Draw!";
            Info.classList.add("large");
            GameContainer.classList.add("end");
            newGame.classList.add("big");
            return;
        }
        
        //Ready Next Player
        p++;
        if(!(p <= players-1)) {
            p = 1;
        }
        Info.textContent = "Player " + p + "'s Turn";
    }
}

function over()
{
    for(r=0; r<rows; r++) {
        for(c=0; c<cols-konnect+1; c++) {
        
            for(k=0; k<konnect; k++)
                if(board[r][c+k].textContent != player[p])
                    break;
            if(k==konnect)
                return true; //Horizontally
        }
    }
    
    for(r=0; r<rows-konnect+1; r++) {
        for(c=0; c<cols; c++) {
    
            for(k=0; k<konnect; k++)
                if(board[r+k][c].textContent != player[p])
                    break;
            if(k==konnect)
                return true; //Vertically
        }
    }
    
    for(r=0; r<rows-konnect+1; r++) {
        for(c=0; c<cols-konnect+1; c++) {
    
            for(k=0; k<konnect; k++)
                if(board[r+k][c+k].textContent != player[p])
                    break;
            if(k==konnect)
                return true; //Diagonally-Right
        }
    }
    
    
    for(r=0; r<rows-konnect+1; r++) {
        for(c=cols-1; c>konnect-2; c--) {
    
            for(k=0; k<konnect; k++)
                if(board[r+k][c-k].textContent != player[p])
                    break;
            if(k==konnect)
                return true; //Diagonally-Left
        }
    }
    
    return false;
}

function draw()
{
	for(r=0; r<rows; r++)
		for(c=0; c<cols; c++)
			if(board[r][c].textContent == emp) //false if even one space is empty
				return false;
	return true;
}

function MuteMusic()
{
    if(GameMusic.muted == true || ConfigureMusic.muted == true) {
        GameMusic.muted = false;
        ConfigureMusic.muted = false;
        SpeakerMuted.style.display = "none";
        Speaker.style.display = "block";
    }
    else {
        if(GameMusic.muted == false && ConfigureMusic.muted == false) {
            ConfigureMusic.play();
        }

        GameMusic.muted = true;
        ConfigureMusic.muted = true;
        SpeakerMuted.style.display = "block";
        Speaker.style.display = "none";
    }
}
